### PR TITLE
NIAD-748: SCR X-Correlation-ID UUID check

### PR DESCRIPTION
**Pull request by Broom94**
## Summary
* Routine Change
* :exclamation: Breaking Change
* :robot: Operational or Infrastructure Change
* :sparkles: New Feature
* :warning: Potential issues that might be caused by this change

Add any other relevant notes or explanations here. **Remove this line if you have nothing to add.**

## Reviews Required
* [x] Dev
* [ ] Test
* [ ] Tech Author
* [ ] Product Owner

## Review Checklist
:information_source: This section is to be filled in by the **reviewer**.

* [ ] I have reviewed the changes in this PR and they fill all or part of the acceptance criteria of the ticket, and the code is in a mergeable state.
* [ ] If there were infrastructure, operational, or build changes, I have made sure there is sufficient evidence that the changes will work.
* [ ] I have ensured the changelog has been updated by the submitter, if necessary.

### CODE REVIEW

File: docker/service/src/main/java/uk/nhs/adaptors/scr/config/ConversationIdFilter.java [hunk 1]
```diff
@@ -40,15 +42,15 @@ protected void doFilterInternal(final HttpServletRequest request, final HttpServ
         }
     }
 
-    public void applyCorrelationId(String id) {
-        MDC.put(MDC_KEY, id);
-    }
-
     public String getRandomCorrelationId() {
         return UUID.randomUUID().toString().toUpperCase().replace("-", "");
```
Comment (bartek-sarul) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/config/ConversationIdFilter.java:46]: I think there is no need to remove dashes
https://digital.nhs.uk/developer/api-catalogue/personal-demographics-service-fhir#api-Relatedperson-getRelatedPeople

File: docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.util.StringUtils;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import uk.nhs.adaptors.scr.components.FhirParser;
+import uk.nhs.adaptors.scr.exceptions.NHSCodings;
+
+public class ExceptionHandlerFilter extends OncePerRequestFilter {
+
+    private static final String HEADER_NAME = "X-Correlation-ID";
```
Comment (bartek-sarul) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java:24]: We should already have this header somewhere as const

File: docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    private static final String UUID_REGEX = "^[0-9a-fA-F]{8}+-[0-9a-fA-F]{4}+-[0-9a-fA-F]{4}+-[0-9a-fA-F]{4}+-[0-9a-fA-F]{12}+$";
+
+    @Override
+    public void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain chain) throws ServletException,
+        IOException {
```
Comment (bartek-sarul) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java:29]: Not a big thing but try to keep same signature parts on the same line. Move "throws..." to new line, or IOException to previous

File: docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        var token = request.getHeader(HEADER_NAME);
+        if (!StringUtils.isEmpty(token) && !token.matches(UUID_REGEX)) {
```
Comment (bartek-sarul) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java:31]: `!StringUtils.isEmpty(token)` can be `StringUtils.isNotEmpty(token)`

File: docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            FhirParser fhirParser = new FhirParser();
+
+            response.setStatus(HttpStatus.BAD_REQUEST.value());
+            response.setContentType(APPLICATION_FHIR_JSON_VALUE);
+
+            var operationOutcome = new OperationOutcome();
+            operationOutcome.addIssue()
+                .setCode(OperationOutcome.IssueType.EXCEPTION)
+                .setSeverity(OperationOutcome.IssueSeverity.ERROR)
+                .setDiagnostics("Invalid " + HEADER_NAME)
```
Comment (bartek-sarul) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java:41]: You can add more information here. Like `Invalid X-Correlation-ID header. Should be an UUIDv4 matching "<regex here>"`

File: docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                .setDetails(new CodeableConcept().addCoding(NHSCodings.BAD_REQUEST.asCoding()));
+
+            response.getWriter().write(fhirParser.encodeResource(APPLICATION_FHIR_JSON, operationOutcome));
```
Comment (bartek-sarul) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java:44]: Will this be json or xml?

File: docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+    private static final String HEADER_NAME = "X-Correlation-ID";
+    private static final String UUID_REGEX = "^[0-9a-fA-F]{8}+-[0-9a-fA-F]{4}+-[0-9a-fA-F]{4}+-[0-9a-fA-F]{4}+-[0-9a-fA-F]{12}+$";
+
+    @Override
+    public void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain chain) throws ServletException,
+        IOException {
+        var token = request.getHeader(HEADER_NAME);
+        if (!StringUtils.isEmpty(token) && !token.matches(UUID_REGEX)) {
+            FhirParser fhirParser = new FhirParser();
```
Comment (bartek-sarul) [inline docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java:32]: Is it possible to inject `FhirParser` ? I think we have a bean regiesterd

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

File: docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java (hunk 6 above)
Comment (Broom94) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java:44]: it returns application/fhir+json

File: docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java (hunk 2 above)
Comment (Broom94) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java:24]: we had the const in another filter class (conversationIdFilter), so i made a new const dir with httpheaders class with both classes using it

File: docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java (hunk 6 above)
Comment (bartek-sarul) [reply docker/service/src/main/java/uk/nhs/adaptors/scr/config/ExceptionHandlerFilter.java:49]: I'm blind. I didn't see that APPLICATION_FHIR_JSON

**Review by bartek-sarul — APPROVED** [review]

**Pull request merged**